File: Data Structures & Algorithms/validate-parentheses/submission-1.py
# class Solution:
#     def isValid(self, s: str) -> bool:
#         stack = []
#         for ch in s:
#             if ch == "(":
#                 stack.append(")")
#             elif ch == "[":
#                 stack.append("]")
#             elif ch == "{":
#                 stack.append("}")
#             elif not len(stack) or stack[-1] != ch:
#                 return False
#             else:
#                 stack.pop()
        
#         return len(stack) == 0

class Solution:
    def isValid(self, s: str) -> bool:
        stack = []
        map = {
            "(": ")",
            "{": "}",
            "[": "]"
        }
        for ch in s:
            if ch in map:
                stack.append(ch)
                continue
            
            if not stack:
                return False
            
            if map[stack.pop()] != ch:
                return False
        
        return len(stack) == 0

File: Data Structures & Algorithms/evaluate-reverse-polish-notation/submission-1.java
class Solution {
    public int evalRPN(String[] tokens) {
        Deque<Integer> stack = new ArrayDeque<>();
        for (String token : tokens) {
            if (isOperator(token)) {
                int right = stack.pop();
                int left = stack.pop();
                if (token.equals("+")) {
                    stack.push(left + right);
                } else if (token.equals("-")) {
                    stack.push(left - right);
                } else if (token.equals("*")) {
                    stack.push(left * right);
                } else if (token.equals("/")) {
                    stack.push(left / right);
                }
            } else {
                stack.push(Integer.parseInt(token));
            }
        }
        return stack.pop();
    }

    private boolean isOperator(String token) {
        return token.equals("+") ||
               token.equals("-") ||
               token.equals("*") ||
               token.equals("/");
    }
}
